runtime/near-vm-runner/src/preload.rs

```diff
@@ -0,0 +1,136 @@
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+
+use threadpool::ThreadPool;
+
+use near_primitives::hash::CryptoHash;
+use near_primitives::runtime::fees::RuntimeFeesConfig;
+use near_primitives::types::CompiledContractCache;
+use near_vm_errors::VMError;
+use near_vm_logic::profile::ProfileData;
+use near_vm_logic::types::PromiseResult;
+use near_vm_logic::{External, ProtocolVersion, VMConfig, VMContext, VMKind, VMOutcome};
+
+use crate::cache;
+use crate::preload::VMModule::Wasmer0;
+use crate::wasmer_runner::run_wasmer_module;
+
+enum VMModule {
+    Wasmer0(wasmer_runtime::Module),
+}
+
+struct VMCallData {
+    result: Result<VMModule, VMError>,
+}
+
+struct CallInner {
+    rx: Receiver<VMCallData>,
+}
+
+// TODO: consider using https://crates.io/crates/scoped_threadpool and not clone the data.
+#[derive(Clone)]
+pub struct ContractCallPrepareRequest {
+    pub code_hash: CryptoHash,
+    pub code: Vec<u8>,
+    pub vm_config: VMConfig,
+    pub cache: Option<Arc<dyn CompiledContractCache>>,
+}
+
+#[derive(Clone)]
+pub struct ContractCallPrepareResult {
+    pub handle: Option<usize>,
+    pub error: Option<VMError>,
+}
+
+pub struct ContractCaller {
+    pool: ThreadPool,
+    prepared: Vec<CallInner>,
+}
+
+impl ContractCaller {
+    pub fn new(num_threads: usize) -> ContractCaller {
+        ContractCaller { pool: ThreadPool::new(num_threads), prepared: Vec::new() }
+    }
+
+    pub fn preload<'a>(
+        self: &mut ContractCaller,
+        requests: Vec<ContractCallPrepareRequest>,
+        vm_kind: VMKind,
+    ) -> Vec<ContractCallPrepareResult> {
+        let mut result: Vec<ContractCallPrepareResult> = Vec::new();
+        for request in requests {
+            let index = self.prepared.len();
+            let (tx, rx) = channel();
+            self.prepared.push(CallInner { rx });
+            let copy_request = request.clone();
+            let tx = tx.clone();
+            self.pool.execute(move || {
+                prepare_in_thread(copy_request, vm_kind, tx);
+            });
+            result.push(ContractCallPrepareResult { handle: Some(index), error: None });
+        }
+        result
+    }
+
+    pub fn run_preloaded<'a>(
+        self: &mut ContractCaller,
+        prepared: &ContractCallPrepareResult,
+        method_name: &str,
+        ext: &mut dyn External,
+        context: VMContext,
+        vm_config: &'a VMConfig,
+        fees_config: &'a RuntimeFeesConfig,
+        promise_results: &'a [PromiseResult],
+        current_protocol_version: ProtocolVersion,
+        profile: ProfileData,
+    ) -> (Option<VMOutcome>, Option<VMError>) {
+        match &prepared.error {
+            Some(error) => return (None, Some(error.clone())),
```

Comment:
Looks like this would never be error because it's constructed by:
```
             result.push(ContractCallPrepareResult { handle: Some(index), error: None });
```
Prepare error would happen here:
```
     tx.send(VMCallData { result }).unwrap();
```
and are saved in self.prepared (`Vec<CallInner>`). so i think to process error, this should refer to `self.prepared[prepared.handle]`?

Reply:
Indeed, made it somewhat simpler.